crates/sparrow-arrow/src/hasher.rs

```diff
@@ -0,0 +1,322 @@
+            if !array.is_null(i) {
+                *hash = array.value(i).hash_one(random_state);
+            }
+        }
+    }
+}
+
+/// Hash the values in a dictionary array
+fn hash_dictionary<K: ArrowDictionaryKeyType>(
+    array: &DictionaryArray<K>,
+    random_state: &RandomState,
+    hashes_buffer: &mut [u64],
+    multi_col: bool,
+) -> error_stack::Result<(), Error> {
+    // Hash each dictionary value once, and then use that computed
+    // hash for each key value to avoid a potentially expensive
+    // redundant hashing for large dictionary elements (e.g. strings)
+    let values = array.values().clone();
+    let mut dict_hashes = vec![0; values.len()];
+    create_hashes(&[values], random_state, &mut dict_hashes, false)?;
+
+    // combine hash for each index in values
+    if multi_col {
+        for (hash, key) in hashes_buffer.iter_mut().zip(array.keys().iter()) {
+            if let Some(key) = key {
+                *hash = combine_hashes(dict_hashes[key.as_usize()], *hash)
+            } // no update for Null, consistent with other hashes
+        }
+    } else {
+        for (hash, key) in hashes_buffer.iter_mut().zip(array.keys().iter()) {
+            if let Some(key) = key {
+                *hash = dict_hashes[key.as_usize()]
+            } // no update for Null, consistent with other hashes
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use arrow::array::{StringArray, UInt64Array};
+    use arrow_array::StructArray;
+    use arrow_schema::{Field, Fields};
+
+    use super::*;
+
+    #[test]
+    fn test_hash_uint64() {
+        let mut hasher = Hasher::default();
+        let array = UInt64Array::from(vec![Some(5), None, Some(8), Some(5), None]);
+        let array: &dyn Array = &array;
+
+        let hashes = hasher.hash_array(array).unwrap();
+        assert_eq!(hashes[0], hashes[3]);
+        assert_eq!(hashes[1], hashes[4]);
+        assert_ne!(hashes[0], hashes[1]);
+        assert_ne!(hashes[0], hashes[2]);
+    }
+
+    #[test]
+    fn test_hash_string() {
+        let array = StringArray::from(vec![
+            Some("hello"),
+            None,
+            Some("world"),
+            Some("hello"),
+            None,
+        ]);
+        let mut hasher = Hasher::default();
+        let array: &dyn Array = &array;
+
+        let hashes = hasher.hash_array(array).unwrap();
+        assert_eq!(hashes[0], hashes[3]);
+        assert_eq!(hashes[1], hashes[4]);
+        assert_ne!(hashes[0], hashes[1]);
+        assert_ne!(hashes[0], hashes[2]);
+    }
+
+    #[test]
+    fn test_hash_string_stability() {
+        let mut hasher = Hasher::default();
+        let array = StringArray::from(vec![Some("hello"), None, Some("world")]);
+        let array: &dyn Array = &array;
+
+        let hashes = hasher.hash_array(array).unwrap();
+        assert_eq!(hashes, &[1472103086483932002, 0, 8057155968893317866]);
```

Comment:
Ideally, I'd delete the old hash code and just always use the `hasher`. We can then move to figuring out how to best manage the "hasher" to share state (I think ideally, we'd have a separate hasher for each thread, or a pool of allocated hashers, etc.).

Unfortunately, since the hashing is slightly different (hashing to `0` when null) this isn't immediately possible.

Reply:
I could make the `hash` kernel just use a thread-local `Hasher`. That would change the hashing a bit -- but that is likely OK. Thoughts on whether I should do that in this PR so that we only have one way of hashing things (and pick-up support for hashing structs)?